File: 1001-2000/1723-find-minimum-time-to-finish-all-jobs/1723-find-minimum-time-to-finish-all-jobs.java
class Solution {
   int[] cap;
  int[] jobs;

  public int minimumTimeRequired(int[] jobs, int k) {
    this.jobs = jobs;
    sort(jobs);
    int n = jobs.length;
    int left = jobs[n - 1];
    int right = jobs[n - 1] * n;
    cap = new int[k];
    while (left < right) {
      int mid = (right + left) >> 1;
      Arrays.fill(cap, mid);
      if (dfs(n - 1, k, mid)) {
        right = mid;
      } else {
        left = mid + 1;
      }
    }
    return left;
  }

  private boolean dfs(int i, int k, int x) {
    if (i == -1) {
      return true;
    }
    for (int j = 0; j < k; j++) {
      if (cap[j] >= jobs[i]) {
        cap[j] -= jobs[i];
        if (dfs(i - 1, k, x))
          return true;

        cap[j] += jobs[i];
      }
      if (cap[j] == x)
        break;

    }
    return false;
  }
  void sort(int[] nums) {
      for (int i = 1; i < nums.length; i++) {
        int j = i;
        while (j > 0 && nums[j - 1] > nums[j])
          swap(nums, j - 1, j--);
      }
    }

    void swap(int[] n, int i, int j) {
      int t = n[i];
      n[i] = n[j];
      n[j] = t;
    }
}

File: 1001-2000/1723-find-minimum-time-to-finish-all-jobs/1723-find-minimum-time-to-finish-all-jobs.cpp
// TC: O(K^N ∗log(Sum))

class Solution {
public:
    bool possible(int jobIdx, vector<int> &workersTime, int maxTimeForEachWorker, int workers, vector<int>& jobs) {
        if(jobIdx == jobs.size()) return true;
        bool poss = false;
        for(int worker = 0; worker < workers; worker++) {
            if(workersTime[worker] + jobs[jobIdx] > maxTimeForEachWorker) continue;
            workersTime[worker] += jobs[jobIdx];
            poss |= possible(jobIdx + 1, workersTime, maxTimeForEachWorker, workers, jobs);
            if(poss) return poss;
            workersTime[worker] -= jobs[jobIdx];
            if (workersTime[worker] == 0) break;
        }
        return poss;
    }

    bool check(int maxTimeForEachWorker, int workers, vector<int>& jobs) {
        vector<int> workersTime(workers, 0); // currently no job is assigned to any worker
        return possible(0, workersTime, maxTimeForEachWorker, workers, jobs);
    }

    int minimumTimeRequired(vector<int>& jobs, int k) {
        sort(jobs.rbegin(), jobs.rend());
        int low = 0, high = accumulate(jobs.begin(), jobs.end(), 0), ans = -1;
        while(low <= high) {
            int mid = (low + high) / 2;
            if(check(mid, k, jobs)) {
                ans = mid;
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }
        return ans;
    }
};